File: atmibroker-xatmi/src/main/cpp/AtmiBroker_ServerImpl.cxx
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2008, Red Hat Middleware LLC, and others contributors as indicated
 * by the @authors tag. All rights reserved.
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
/*
 * BREAKTHRUIT PROPRIETARY - NOT TO BE DISCLOSED OUTSIDE BREAKTHRUIT, LLC.
 */
// copyright 2006, 2008 BreakThruIT

//
// Servant which implements the AtmiBroker::Server interface.
//

#ifdef TAO_COMP
#include <orbsvcs/CosNamingS.h>
#endif

#include <string>
#include <queue>
#include "AtmiBroker_ServerImpl.h"
#include "AtmiBroker.h"
#include "AtmiBrokerPoaFac.h"
#include "AtmiBrokerEnv.h"
#include "log4cxx/logger.h"
#include "EndpointQueue.h"

log4cxx::LoggerPtr loggerAtmiBroker_ServerImpl(log4cxx::Logger::getLogger("AtmiBroker_ServerImpl"));

// AtmiBroker_ServerImpl constructor
//
// Note: since we use virtual inheritance, we must include an
// initialiser for all the virtual base class constructors that
// require arguments, even those that we inherit indirectly.
//
AtmiBroker_ServerImpl::AtmiBroker_ServerImpl(CONNECTION* connection, PortableServer::POA_ptr poa) {
	AtmiBrokerServerXml aAtmiBrokerServerXml;
	aAtmiBrokerServerXml.parseXmlDescriptor(&serverInfo, (char*) "SERVER.xml");
	serverName = server;
	this->connection = connection;
	this->poa = poa;
}

// ~AtmiBroker_ServerImpl destructor.
//
AtmiBroker_ServerImpl::~AtmiBroker_ServerImpl() {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "destructor ");
	serviceData.clear();
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "deleted service array ");
}

// server_init() -- Implements IDL operation "AtmiBroker::Server::server_init".
//
CORBA::Short AtmiBroker_ServerImpl::server_init() throw (CORBA::SystemException ) {
	int toReturn = 0;
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "server_init(): called.");

	PortableServer::ObjectId_var oid = PortableServer::string_to_ObjectId(server);
	poa->activate_object_with_id(oid, this);
	CORBA::Object_var tmp_ref = poa->create_reference_with_id(oid, "IDL:AtmiBroker/Server:1.0");

	CosNaming::Name * name = ((CosNaming::NamingContextExt_ptr) connection->default_ctx)->to_name(serverName);
	((CosNaming::NamingContext_ptr) connection->name_ctx)->bind(*name, tmp_ref);
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "server_init(): finished.");
	return toReturn;
}

// server_done() -- Implements IDL operation "AtmiBroker::Server::server_done".
//
void AtmiBroker_ServerImpl::server_done() throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "server_done()");

	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "unadvertise " << serverName);

	CosNaming::Name* name = ((CosNaming::NamingContextExt_ptr) connection->default_ctx)->to_name(serverName);
	((CosNaming::NamingContext_ptr) connection->name_ctx)->unbind(*name);

	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "unadvertised " << serverName);

	for (unsigned int i = 0; i < serverInfo.serviceNames.size(); i++) {
		char* svcname = (char*) serverInfo.serviceNames[i].c_str();
		if (isAdvertised(svcname)) {
			unadvertiseService(svcname);
		}
	}
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "server_done(): returning.");
}

char *
AtmiBroker_ServerImpl::getServerName() {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "getServerName ");
	return serverName;
}

bool AtmiBroker_ServerImpl::advertiseService(char * serviceName, void(*func)(TPSVCINFO *)) {
	if (!serviceName || strlen(serviceName) == 0) {
		tperrno = TPEINVAL;
		return false;
	}

	bool found = false;
	for (unsigned int i = 0; i < serverInfo.serviceNames.size(); i++) {
		if (strncmp(serverInfo.serviceNames[i].c_str(), serviceName, XATMI_SERVICE_NAME_LENGTH) == 0) {
			found = true;
			break;
		}
	}
	if (!found) {
		tperrno = TPELIMIT;
		return false;
	}
	void (*serviceFunction)(TPSVCINFO*) = getServiceMethod(serviceName);
	if (serviceFunction != NULL) {
		if (serviceFunction == func) {
			return true;
		} else {
			tperrno = TPEMATCH;
			return false;
		}
	}

	bool toReturn = false;

	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "advertiseService(): " << serviceName);

	// create reference for Service Queue and cache
	try {
		LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "create_service_queue: " << serviceName);

		// create Poa for Service Queue
		LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "create_service_queue_poa: " << serviceName);
		AtmiBrokerPoaFac* poaFactory = ((AtmiBrokerPoaFac*) connection->poaFactory);
		PortableServer::POA_ptr aFactoryPoaPtr = poaFactory->createServicePoa((CORBA::ORB_ptr) connection->orbRef, serviceName, poa, (PortableServer::POAManager_ptr) connection->root_poa_manager);
		LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "created create_service_factory_poa: " << serviceName);

		Destination* destination = ::create_service_queue(serverConnection, aFactoryPoaPtr, serviceName);
		ServiceQueue *tmp_factory_servant = new ServiceQueue(serverConnection, destination, serviceName, func);
		LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) " tmp_factory_servant " << tmp_factory_servant);

		addServiceQueue(serviceName, tmp_factory_servant, func);
		LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "created ServiceQueue " << serviceName);

	} catch (...) {
		LOG4CXX_ERROR(loggerAtmiBroker_ServerImpl, (char*) "service has already been advertised, however it appears to be by a different server (possibly with the same name), which is strange... " << serviceName);
		tperrno = TPEMATCH;
		return false;
	}

	advertisedServices.push_back(serviceName);
	LOG4CXX_INFO(loggerAtmiBroker_ServerImpl, (char*) "advertised service " << serviceName);
	toReturn = true;
	return toReturn;
}

// TODO CLEANUP SERVICE POA AND SERVANT CACHE
void AtmiBroker_ServerImpl::unadvertiseService(char * serviceName) {
	for (std::vector<char*>::iterator i = advertisedServices.begin(); i != advertisedServices.end(); i++) {
		if (strcmp(serviceName, (*i)) == 0) {
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "remove_service_queue: " << serviceName);
			CosNaming::Name * name = ((CosNaming::NamingContextExt_ptr) connection->default_ctx)->to_name(serviceName);
			((CosNaming::NamingContext_ptr) connection->name_ctx)->unbind(*name);

			ServiceQueue* toDelete = removeServiceQueue(serviceName);
			EndpointQueue* queue = dynamic_cast<EndpointQueue*> (toDelete->getDestination());
			PortableServer::POA_ptr poa = (PortableServer::POA_ptr) queue->getPoa();
			delete toDelete;
			poa->destroy(true, true);
			poa = NULL;
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "service Queue removed" << serviceName);
			advertisedServices.erase(i);
			LOG4CXX_INFO(loggerAtmiBroker_ServerImpl, (char*) "unadvertised service " << serviceName);
			break;
		}
	}
}

bool AtmiBroker_ServerImpl::isAdvertised(char * serviceName) {
	bool toReturn = false;
	for (std::vector<char*>::iterator i = advertisedServices.begin(); i != advertisedServices.end(); i++) {
		if (strcmp(serviceName, (*i)) == 0) {
			toReturn = true;
		}
	}
	return toReturn;
}

// get_server_info() -- Implements IDL operation "AtmiBroker::Server::get_server_info".
//
AtmiBroker::ServerInfo*
AtmiBroker_ServerImpl::get_server_info() throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "get_server_info()");

	AtmiBroker::ServerInfo_var aServerInfo = new AtmiBroker::ServerInfo();

	aServerInfo->maxChannels = serverInfo.maxChannels;
	aServerInfo->maxSuppliers = serverInfo.maxSuppliers;
	aServerInfo->maxConsumers = serverInfo.maxConsumers;
	aServerInfo->logLevel = serverInfo.logLevel;
	aServerInfo->securityType = CORBA::string_dup(serverInfo.securityType.c_str());
	aServerInfo->orbType = CORBA::string_dup(serverInfo.orbType.c_str());
	aServerInfo->queueSpaceName = CORBA::string_dup(queue_name);

	//aServerInfo->serviceNames.length(serverInfo.serviceNames.size());
	std::queue<AtmiBroker::octetSeq *> returnData;
	for (unsigned int i = 0; i < serverInfo.serviceNames.size(); i++) {
		std::string serviceName = serverInfo.serviceNames[i];
		//AtmiBroker::octetSeq * aOctetSeq =
		returnData.push(new AtmiBroker::octetSeq(serviceName.length(), serviceName.length(), (unsigned char *) serviceName.c_str(), true));//*aOctetSeq));
	}
	AtmiBroker::ServiceNameSeq_var aOctetSeq = new AtmiBroker::ServiceNameSeq(serverInfo.serviceNames.size(), serverInfo.serviceNames.size(), returnData.front(), true);
	aServerInfo->serviceNames = aOctetSeq;

	return aServerInfo._retn();
}

// get_all_service_info() -- Implements IDL operation "AtmiBroker::Server::get_all_service_info".
//
AtmiBroker::ServiceInfoSeq*
AtmiBroker_ServerImpl::get_all_service_info() throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "get_all_service_info()");

	AtmiBroker::ServiceInfoSeq_var aServiceInfoSeq = new AtmiBroker::ServiceInfoSeq();
	aServiceInfoSeq->length(serverInfo.serviceNames.size());

	for (unsigned int i = 0; i < serverInfo.serviceNames.size(); i++) {
		SVCINFO svcInfo = getServiceQueue((char*) serverInfo.serviceNames[i].c_str())->get_service_info();
		AtmiBroker::ServiceInfo_var aServiceInfo = new AtmiBroker::ServiceInfo();
		aServiceInfo->serviceName = svcInfo.serviceName;
		aServiceInfo->poolSize = svcInfo.poolSize;
		aServiceInfo->securityType = svcInfo.securityType;
		// TODO(*aServiceInfoSeq)[i] = aServiceInfo._retn();
	}
	return aServiceInfoSeq._retn();
}

// get_environment_variable_info() -- Implements IDL operation "AtmiBroker::Server::get_environment_variable_info".
//
AtmiBroker::EnvVariableInfoSeq*
AtmiBroker_ServerImpl::get_environment_variable_info() throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "get_environment_variable_info()");

	std::vector<envVar_t> & aEnvVarInfoSeq = AtmiBrokerEnv::get_instance()->getEnvVariableInfoSeq();

	AtmiBroker::EnvVariableInfoSeq* aEnvVariableInfoSeqPtr = new AtmiBroker::EnvVariableInfoSeq();

	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "get_environment_variable_info() setting length to " << aEnvVarInfoSeq.size());
	aEnvVariableInfoSeqPtr->length(aEnvVarInfoSeq.size());
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "get_environment_variable_info() set length to " << aEnvVariableInfoSeqPtr->length());
	int j = 0;
	for (std::vector<envVar_t>::iterator i = aEnvVarInfoSeq.begin(); i != aEnvVarInfoSeq.end(); i++) {
		LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "get_environment_variable_info() env name " << (*i).name);
		(*aEnvVariableInfoSeqPtr)[j].name = strdup((*i).name);
		(*aEnvVariableInfoSeqPtr)[j].value = strdup((*i).value);
		j++;
	}
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "get_environment_variable_info() returning ");
	return aEnvVariableInfoSeqPtr;
}

// set_server_descriptor() -- Implements IDL operation "AtmiBroker::Server::set_server_descriptor".
//
void AtmiBroker_ServerImpl::set_server_descriptor(const char* xml_descriptor) throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_server_descriptor() " << xml_descriptor);

	std::string serverFileName = "server.xml";

	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_server_descriptor() file name " << serverFileName.c_str());
	FILE* aTempFile = fopen(serverFileName.c_str(), "w");
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_server_descriptor() file opened " << aTempFile);

	fputs(xml_descriptor, aTempFile);

	fflush(aTempFile);
	fclose(aTempFile);
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_server_descriptor() file written and closed ");
}

// set_service_descriptor() -- Implements IDL operation "AtmiBroker::Server::set_service_descriptor".
//
void AtmiBroker_ServerImpl::set_service_descriptor(const char* service_name, const char* xml_descriptor) throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_service_descriptor() %s " << service_name << " " << xml_descriptor);

	std::string serverFileName = "server.xml";

	FILE* aTempFile = fopen(serverFileName.c_str(), "w");
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_service_descriptor()  file opened " << " " << serverFileName.c_str() << " " << aTempFile);

	fputs(xml_descriptor, aTempFile);

	fflush(aTempFile);
	fclose(aTempFile);
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_service_descriptor()  file written and closed ");
}

// set_environment_descriptor() -- Implements IDL operation "AtmiBroker::Server::set_environment_descriptor".
//
void AtmiBroker_ServerImpl::set_environment_descriptor(const char* xml_descriptor) throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_environment_descriptor() " << xml_descriptor);

	FILE* aTempFile = fopen("Environment.xml", "w");
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_environment_descriptor() file Environment.xml opened " << aTempFile);

	fputs(xml_descriptor, aTempFile);

	fflush(aTempFile);
	fclose(aTempFile);
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "set_environment_descriptor() file written and closed ");
}

// stop_service() -- Implements IDL operation "AtmiBroker::Server::stop_service".
//
void AtmiBroker_ServerImpl::stop_service(const char* service_name) throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "stop_service() " << service_name);

	try {
		for (unsigned int i = 0; i < serverInfo.serviceNames.size(); i++) {
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "stop_service() next service " << (const char*) serverInfo.serviceNames[i].c_str());
			if (strcmp(service_name, (const char*) serverInfo.serviceNames[i].c_str()) == 0) {
				LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "stop_service() found matching service calling tpunadvertise" << service_name);
#ifndef VBC_COMP
				tpunadvertise((char*) serverInfo.serviceNames[i].c_str());
#else
				tpunadvertise((char*)serverInfo.serviceNames[i]);
#endif
				LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "stop_service() called tpunadvertise");
				return;
			}
		}
		LOG4CXX_ERROR(loggerAtmiBroker_ServerImpl, (char*) "stop_service() DID NOT find matching service " << service_name);
	} catch (CORBA::Exception &e) {
		LOG4CXX_ERROR(loggerAtmiBroker_ServerImpl, (char*) "stop_service() Could not stop service - exception: " << e._name());
	}
}

// start_service() -- Implements IDL operation "AtmiBroker::Server::start_service".
//
void AtmiBroker_ServerImpl::start_service(const char* service_name) throw (CORBA::SystemException ) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "start_service() " << service_name);

	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "start_service()  stopping service first ");
	stop_service(service_name);
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "start_service()  stopped service");

	try {
		for (unsigned int i = 0; i < serverInfo.serviceNames.size(); i++) {
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "start_service()  next service " << (const char*) serverInfo.serviceNames[i].c_str());
			if (strcmp(service_name, (const char*) serverInfo.serviceNames[i].c_str()) == 0) {
				LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "start_service()  found matching service calling tpadvertise" << service_name);
#ifndef VBC_COMP
				tpadvertise((char*) serverInfo.serviceNames[i].c_str(), NULL);
#else
				tpadvertise((char*)serverInfo.serviceNames[i], NULL);
#endif
				LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "start_service()  called tpadvertise");
				return;
			}
		}
		LOG4CXX_ERROR(loggerAtmiBroker_ServerImpl, (char*) "start_service()  DID NOT find matching service " << service_name);
	} catch (CORBA::Exception &e) {
		LOG4CXX_ERROR(loggerAtmiBroker_ServerImpl, (char*) "start_service()  Could not start service - exception: " << e._name());
	}
}

ServiceQueue* AtmiBroker_ServerImpl::getServiceQueue(const char * aServiceName) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "getServiceQueue: " << aServiceName);

	for (std::vector<ServiceData>::iterator i = serviceData.begin(); i != serviceData.end(); i++) {
		if (strncmp((*i).serviceName, aServiceName, XATMI_SERVICE_NAME_LENGTH) == 0) {
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "found: " << (char*) (*i).serviceName);
			return (*i).serviceQueue;
		}
	}
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "getServiceQueue out: " << aServiceName);
	return NULL;
}

void AtmiBroker_ServerImpl::addServiceQueue(char*& aServiceName, ServiceQueue*& aFactoryPtr, void(*func)(TPSVCINFO *)) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "addServiceQueue: " << aServiceName);

	ServiceData entry;
	entry.serviceName = aServiceName;
	entry.serviceQueue = aFactoryPtr;
	entry.func = func;
	serviceData.push_back(entry);
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "added: " << (char*) aServiceName);
}

ServiceQueue* AtmiBroker_ServerImpl::removeServiceQueue(const char * aServiceName) {
	LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "removeServiceQueue: " << aServiceName);
	ServiceQueue* toReturn = NULL;
	for (std::vector<ServiceData>::iterator i = serviceData.begin(); i != serviceData.end(); i++) {
		if (strncmp((*i).serviceName, aServiceName, XATMI_SERVICE_NAME_LENGTH) == 0) {
			toReturn = (*i).serviceQueue;
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "removing service " << (char*) (*i).serviceName);
			serviceData.erase(i);
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "removed: " << aServiceName);
			break;
		}
	}
	return toReturn;
}

void (*AtmiBroker_ServerImpl::getServiceMethod(const char * aServiceName))(TPSVCINFO *) {
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "getServiceQueue: " << aServiceName);

			for (std::vector<ServiceData>::iterator i = serviceData.begin(); i != serviceData.end(); i++) {
				if (strncmp((*i).serviceName, aServiceName, XATMI_SERVICE_NAME_LENGTH) == 0) {
					LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "found: " << (char*) (*i).serviceName);
					return (*i).func;
				}
			}
			LOG4CXX_DEBUG(loggerAtmiBroker_ServerImpl, (char*) "getServiceMethod out: " << aServiceName);
			return NULL;
		}

File: jatmibroker-core/src/main/java/org/jboss/blacktie/jatmibroker/core/AtmiBroker_ServerImpl.java
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2008, Red Hat Middleware LLC, and others contributors as indicated
 * by the @authors tag. All rights reserved.
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
package org.jboss.blacktie.jatmibroker.core;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Properties;

import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.omg.CORBA.Object;
import org.omg.CORBA.Policy;
import org.omg.CosNaming.NameComponent;
import org.omg.PortableServer.POA;
import org.omg.PortableServer.ThreadPolicyValue;

import AtmiBroker.ClientInfo;
import AtmiBroker.EnvVariableInfo;
import AtmiBroker.Server;
import AtmiBroker.ServerInfo;
import AtmiBroker.ServerPOA;
import AtmiBroker.ServiceInfo;

public class AtmiBroker_ServerImpl extends ServerPOA {
	private static final Logger log = LogManager.getLogger(AtmiBroker_ServerImpl.class);
	private POA poa;
	private Server me;
	private String serverName;
	private byte[] activate_object;
	private Map<String, AtmiBroker_ServiceFactoryImpl> serviceFactoryList = new HashMap<String, AtmiBroker_ServiceFactoryImpl>();
	private boolean bound;

	public AtmiBroker_ServerImpl(Properties properties) throws JAtmiBrokerException {
		String companyName = properties.getProperty("blacktie.company.name");
		String serverName = properties.getProperty("blacktie.server.name");
		int numberOfOrbArgs = Integer.parseInt(properties.getProperty("blacktie.orb.args"));
		List<String> orbArgs = new ArrayList<String>(numberOfOrbArgs);
		for (int i = 1; i <= numberOfOrbArgs; i++) {
			orbArgs.add(properties.getProperty("blacktie.orb.arg." + i));
		}
		String[] args = orbArgs.toArray(new String[] {});
		try {
			AtmiBrokerServerImpl.ConnectToORB(args, companyName);
		} catch (Throwable t) {
			throw new JAtmiBrokerException("Could not connect to orb for company name: " + companyName, t);
		}
		this.serverName = serverName;
		Policy[] policiesArray = new Policy[1];
		List<Policy> policies = new ArrayList<Policy>(1);
		policies.add(AtmiBrokerServerImpl.root_poa.create_thread_policy(ThreadPolicyValue.ORB_CTRL_MODEL));
		policies.toArray(policiesArray);
		try {
			this.poa = AtmiBrokerServerImpl.root_poa.create_POA(serverName, AtmiBrokerServerImpl.root_poa.the_POAManager(), policiesArray);
		} catch (Throwable t) {
			try {
				this.poa = AtmiBrokerServerImpl.root_poa.find_POA(serverName, true);
			} catch (Throwable t2) {
				throw new JAtmiBrokerException("Could not find poa", t2);
			}
		}
	}

	public void createAtmiBroker_ServiceFactoryImpl(String serviceName, int servantCacheSize, Class callback, AtmiBroker_CallbackConverter atmiBroker_Callback) throws JAtmiBrokerException {
		if (!serviceFactoryList.containsKey(serviceName)) {
			try {
				AtmiBroker_ServiceFactoryImpl atmiBroker_ServiceFactoryImpl = new AtmiBroker_ServiceFactoryImpl(serviceName, servantCacheSize, callback, atmiBroker_Callback);
				serviceFactoryList.put(serviceName, atmiBroker_ServiceFactoryImpl);
				if (bound) {
					atmiBroker_ServiceFactoryImpl.bind();
				}
			} catch (Throwable t) {
				throw new JAtmiBrokerException("Could not create service factory for: " + serviceName, t);
			}
		}
	}

	public void bind() throws JAtmiBrokerException {
		try {
			activate_object = poa.activate_object(this);
			NameComponent[] name = AtmiBrokerServerImpl.nce.to_name(serverName);
			Object servant_to_reference = poa.servant_to_reference(this);
			AtmiBrokerServerImpl.nc.rebind(name, servant_to_reference);
			me = AtmiBroker.ServerHelper.narrow(servant_to_reference);
		} catch (Throwable t) {
			throw new JAtmiBrokerException("Could not bind server", t);
		}
		Iterator<AtmiBroker_ServiceFactoryImpl> iterator = serviceFactoryList.values().iterator();
		while (iterator.hasNext()) {
			iterator.next().bind();
		}
		bound = true;
	}

	public void unbind() throws JAtmiBrokerException {
		Iterator<AtmiBroker_ServiceFactoryImpl> iterator = serviceFactoryList.values().iterator();
		while (iterator.hasNext()) {
			iterator.next().unbind();
		}
		try {
			poa.deactivate_object(activate_object);
		} catch (Throwable t) {
			throw new JAtmiBrokerException("Could not unbind server", t);
		}
		bound = false;
	}

	public void unbind(String serviceName) throws JAtmiBrokerException {
		AtmiBroker_ServiceFactoryImpl atmiBroker_ServiceFactoryImpl = serviceFactoryList.get(serviceName);
		if (atmiBroker_ServiceFactoryImpl != null) {
			atmiBroker_ServiceFactoryImpl.unbind();
		}
	}

	public boolean deregister_client(ClientInfo client_info) {
		log.error("NO-OP deregister_client");
		// TODO Auto-generated method stub
		return false;
	}

	public ServiceInfo[] get_all_service_info() {
		log.error("NO-OP get_all_service_info");
		// TODO Auto-generated method stub
		return null;
	}

	public String get_client_callback(ClientInfo client_info) {
		log.error("NO-OP get_client_callback");
		// TODO Auto-generated method stub
		return null;
	}

	public ClientInfo[] get_client_info() {
		log.error("NO-OP get_client_info");
		// TODO Auto-generated method stub
		return null;
	}

	public EnvVariableInfo[] get_environment_variable_info() {
		log.error("NO-OP get_environment_variable_info");
		// TODO Auto-generated method stub
		return null;
	}

	public int get_queue_log(String queue_name) {
		log.error("NO-OP get_queue_log");
		// TODO Auto-generated method stub
		return 0;
	}

	public ServerInfo get_server_info() {
		log.error("NO-OP get_server_info");
		// TODO Auto-generated method stub
		return null;
	}

	public ServiceInfo get_service_info(String service_name) {
		log.error("NO-OP get_service_info");
		// TODO Auto-generated method stub
		return null;
	}

	public int register_client(ClientInfo client_info) {
		log.error("NO-OP register_client");
		// TODO Auto-generated method stub
		return 0;
	}

	public void server_done() {
		log.error("NO-OP server_done");
		// TODO Auto-generated method stub

	}

	public short server_init() {
		log.error("NO-OP server_init");
		// TODO Auto-generated method stub
		return 0;
	}

	public void set_environment_descriptor(String xml_descriptor) {
		log.error("NO-OP set_environment_descriptor");
		// TODO Auto-generated method stub

	}

	public void set_server_descriptor(String xml_descriptor) {
		log.error("NO-OP set_server_descriptor");
		// TODO Auto-generated method stub

	}

	public void set_service_descriptor(String serviceName, String xml_descriptor) {
		log.error("NO-OP set_service_descriptor");
		// TODO Auto-generated method stub

	}

	public void start_service(String service_name) {
		log.error("NO-OP start_service");
		// TODO Auto-generated method stub

	}

	public void stop_service(String service_name) {
		log.error("NO-OP stop_service");
		// TODO Auto-generated method stub

	}
}
